### PR TITLE
[WIP] Support an asynchronous spawning interface

### PR DESCRIPTION
**Pull request by noxiouz**
It's just a propose of how an asynchronous spawning interface might look. Also process isolation is ported to this intherface.

### CODE REVIEW

File: node/src/node/slave/state/spawning.cpp [hunk 1]
```diff
@@ -111,9 +106,15 @@ spawning_t::spawn(unsigned long timeout) {
 }
 
 void
-spawning_t::on_spawn(std::chrono::high_resolution_clock::time_point start) {
+spawning_t::on_spawn(std::chrono::high_resolution_clock::time_point start, const std::error_code& err, std::unique_ptr<api::handle_t>& handle) {
     std::error_code ec;
     const size_t cancelled = timer.cancel(ec);
+
+    if (err) {
+        COCAINE_LOG_WARNING(slave->log, "failed to spawn a slave %d", err);
```
Comment (3Hren) [inline node/src/node/slave/state/spawning.cpp]: The format is invalid.

File: node/include/cocaine/detail/service/node/slave/state/spawning.hpp [hunk 2]
```diff
@@ -4,6 +4,8 @@
 
 #include "state.hpp"
 
+#include "cocaine/api/isolate.hpp"
```
Comment (3Hren) [inline node/include/cocaine/detail/service/node/slave/state/spawning.hpp]: There is no need for this include, use forwards.

File: node/src/node/slave/state/spawning.cpp [hunk 3]
```diff
@@ -139,12 +139,12 @@ spawning_t::on_spawn(std::chrono::high_resolution_clock::time_point start) {
...
 
 void
-spawning_t::on_timeout(const std::error_code& ec) {
+spawning_t::on_timeout(std::unique_ptr<api::cancellation_t>& cancellation, const std::error_code& ec) {
     if (ec) {
         COCAINE_LOG_TRACE(slave->log, "spawn timer has called its completion handler: cancelled");
     } else {
         COCAINE_LOG_ERROR(slave->log, "unable to spawn slave: timeout");
-
+        cancellation->cancel();
```
Comment (3Hren) [inline node/src/node/slave/state/spawning.cpp]: This won't be cancelled if the slave calls `spawning_t::cancel` before timeout reached.

File: node/src/isolate/process.cpp [hunk 4]
```diff
@@ -399,12 +400,17 @@ process_t::~process_t() {
...
 
-std::unique_ptr<api::handle_t>
-process_t::spawn(const std::string& path, const api::string_map_t& args, const api::string_map_t& environment) {
+std::unique_ptr<api::cancellation_t>
+process_t::spawn(const std::string& path, const api::string_map_t& args, const api::string_map_t& environment, api::spawn_handler_t handler) {
     std::array<int, 2> pipes;
 
     if(::pipe(pipes.data()) != 0) {
-        throw std::system_error(errno, std::system_category(), "unable to create an output pipe");
+        COCAINE_LOG_ERROR(m_log, "unable to create an output pipe: %d", errno);
```
Comment (3Hren) [inline node/src/isolate/process.cpp]: When it's time to format the message an `errno` can be changed for some other reasons than the initial one.

File: node/src/isolate/process.cpp [hunk 5]
```diff
@@ -415,20 +421,29 @@ process_t::spawn(const std::string& path, const api::string_map_t& args, const a
 
     if(pid < 0) {
         std::for_each(pipes.begin(), pipes.end(), ::close);
+        COCAINE_LOG_ERROR(m_log, "unable to fork: %d", errno);
```
Comment (3Hren) [inline node/src/isolate/process.cpp]: And here.

File: node/src/isolate/process.cpp [hunk 6]
```diff
@@ -415,20 +421,29 @@ process_t::spawn(const std::string& path, const api::string_map_t& args, const a
...
 
-        throw std::system_error(errno, std::system_category(), "unable to fork");
+        get_io_service().post(std::bind(handler,
+                                std::error_code(errno, std::system_category()),
```
Comment (3Hren) [inline node/src/isolate/process.cpp]: Check indents please.

File: node/src/node/slave/state/spawning.cpp [hunk 7]
```diff
@@ -144,7 +145,8 @@ spawning_t::on_timeout(const std::error_code& ec) {
         COCAINE_LOG_TRACE(slave->log, "spawn timer has called its completion handler: cancelled");
     } else {
         COCAINE_LOG_ERROR(slave->log, "unable to spawn slave: timeout");
-
         slave->shutdown(error::spawn_timeout);
     }
+
+    cancellation->cancel();
```
Comment (noxiouz) [inline node/src/node/slave/state/spawning.cpp]: I have  to remove it.

**Comment by 3Hren** [comment]
This ain't gonna work until `trace_t` becomes movable again.

**Pull request closed without merging**